### PR TITLE
fix JSON bug with data reading

### PR DESCRIPTION
**Pull request by JGSweets**
Previously: we could not read line separated JSON arrays.
Now we can read:
```
[1, 2]
[2, 3]
[3, 3]
```
becomes a JSONData which is wrapping pandas as:
```
   0   1
0  1   2
1  2   3
2  3   3
```

### CODE REVIEW

File: dataprofiler/data_readers/avro_data.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 """Contains class for saving and loading spreadsheet data."""
 from io import BytesIO, StringIO
-from typing import Any, Dict, List, Optional, Union, cast
+from typing import Any, Dict, List, Optional, Union
```
Comment (JGSweets) [inline dataprofiler/data_readers/avro_data.py:3]: the proper way in the other func for a boolean which checks type is to return a TypeGuard

Comment (JGSweets) [reply dataprofiler/data_readers/avro_data.py:3]: no need to cast now

File: dataprofiler/data_readers/avro_data.py [hunk 2]
```diff
@@ -92,18 +92,12 @@ def is_match(
 
         # get current position of stream
         if data_utils.is_stream_buffer(file_path):
-            file_path = cast(
```
Comment (JGSweets) [inline dataprofiler/data_readers/avro_data.py:95]: static typing, not functional change, removed bc of TypeGuard

File: dataprofiler/data_readers/avro_data.py [hunk 3]
```diff
@@ -92,18 +92,12 @@ def is_match(
...
-                Union[StringIO, BytesIO], file_path
-            )  # guaranteed by is_stream_buffer
             starting_location = file_path.tell()
 
         is_valid_avro = fastavro.is_avro(file_path)
 
         # return to original position in stream
         if data_utils.is_stream_buffer(file_path):
-            file_path = cast(
```
Comment (JGSweets) [inline dataprofiler/data_readers/avro_data.py:104]: static typing, not functional change, removed bc of TypeGuard

File: dataprofiler/data_readers/filepath_or_buffer.py [hunk 4]
```diff
@@ -2,8 +2,10 @@
 from io import BytesIO, StringIO, TextIOWrapper, open
 from typing import IO, Any, Optional, Type, Union, cast
 
+from typing_extensions import TypeGuard
 
-def is_stream_buffer(filepath_or_buffer: Any) -> bool:
+
+def is_stream_buffer(filepath_or_buffer: Any) -> TypeGuard[Union[StringIO, BytesIO]]:
```
Comment (JGSweets) [inline dataprofiler/data_readers/filepath_or_buffer.py:8]: fix to use TypeGuard

File: dataprofiler/data_readers/filepath_or_buffer.py [hunk 5]
```diff
@@ -27,10 +29,10 @@ class FileOrBufferHandler:
...
         self,
         filepath_or_buffer: Union[str, StringIO, BytesIO],
-        open_method: str ="r",
-        encoding: Optional[str]=None,
-        seek_offset: Optional[int]=None,
-        seek_whence: int=0,
+        open_method: str = "r",
+        encoding: Optional[str] = None,
+        seek_offset: Optional[int] = None,
+        seek_whence: int = 0,
```
Comment (JGSweets) [inline dataprofiler/data_readers/filepath_or_buffer.py:35]: formatting

File: dataprofiler/data_readers/filepath_or_buffer.py [hunk 6]
```diff
@@ -51,7 +53,9 @@ def __init__(
         self.seek_offset: Optional[int] = seek_offset
         self.seek_whence: int = seek_whence
         self._encoding: Optional[str] = encoding
-        self.original_type: Union[Type[str], Type[StringIO], Type[BytesIO], Type[IO]] = type(filepath_or_buffer)
+        self.original_type: Union[
+            Type[str], Type[StringIO], Type[BytesIO], Type[IO]
+        ] = type(filepath_or_buffer)
```
Comment (JGSweets) [inline dataprofiler/data_readers/filepath_or_buffer.py:58]: format

File: dataprofiler/data_readers/filepath_or_buffer.py [hunk 7]
```diff
@@ -84,13 +88,17 @@ def __exit__(self, exc_type: Any, exc_value: Any, exc_traceback: Any) -> None:
         """Release resources."""
         # Need to detach buffer if wrapped (i.e. BytesIO opened with 'r')
         if self._is_wrapped:
-            self._filepath_or_buffer = cast(TextIOWrapper, self._filepath_or_buffer) # guaranteed by self._is_wrapped
+            self._filepath_or_buffer = cast(
+                TextIOWrapper, self._filepath_or_buffer
+            )  # guaranteed by self._is_wrapped
```
Comment (JGSweets) [inline dataprofiler/data_readers/filepath_or_buffer.py:93]: format

File: dataprofiler/data_readers/filepath_or_buffer.py [hunk 8]
```diff
@@ -84,13 +88,17 @@ def __exit__(self, exc_type: Any, exc_value: Any, exc_traceback: Any) -> None:
...
             self._filepath_or_buffer = wrapper.buffer
             wrapper.detach()
 
         if isinstance(self._filepath_or_buffer, (StringIO, BytesIO)):
             self._filepath_or_buffer.seek(0)
         else:
-            self._filepath_or_buffer = cast(IO, self._filepath_or_buffer) # can't be str due to conversion in __enter__
+            self._filepath_or_buffer = cast(
+                IO, self._filepath_or_buffer
+            )  # can't be str due to conversion in __enter__
```
Comment (JGSweets) [inline dataprofiler/data_readers/filepath_or_buffer.py:103]: format

File: dataprofiler/data_readers/json_data.py [hunk 9]
```diff
@@ -276,36 +277,35 @@ def _load_data_from_file(self, input_file_path: str) -> List:
...
         :param data: raw data
         :type data: list
         :return: dataframe in record format
         """
-        _data: Union[pd.DataFrame, List]
-        _data = self._get_data_as_df(data)
-        _data = _data.to_dict(orient="records", into=OrderedDict)
-        for i, sample in enumerate(_data):
-            _data[i] = json.dumps(
+        data = self._get_data_as_df(data)
```
Comment (JGSweets) [inline dataprofiler/data_readers/json_data.py:288]: no longer having two variables of the data with `_data`
fixes type acceptance at the top

File: dataprofiler/data_readers/json_data.py [hunk 10]
```diff
@@ -276,36 +277,35 @@ def _load_data_from_file(self, input_file_path: str) -> List:
...
-    def _get_data_as_json(self, data: List) -> List[str]:
+    def _get_data_as_json(self, data: Union[pd.DataFrame, Dict, List]) -> List[str]:
         """
         Extract the data as a json format.
 
         :param data: raw data
         :type data: list
         :return: dataframe in json format
         """
-        _data: Union[pd.DataFrame, List]
```
Comment (JGSweets) [inline dataprofiler/data_readers/json_data.py:304]: no longer having two variables of the data with _data
fixes type acceptance at the top

File: dataprofiler/data_readers/json_data.py [hunk 11]
```diff
@@ -349,6 +349,8 @@ def _convert_flat_to_nested_cols(cls, dic: Dict, separator: str = ".") -> Dict:
         :return:
         """
         for key in list(dic.keys()):
+            if not isinstance(key, str):
```
Comment (JGSweets) [inline dataprofiler/data_readers/json_data.py:352]: fix for allowing the `[1]` format in json reading

File: dataprofiler/data_readers/json_data.py [hunk 12]
```diff
@@ -392,14 +394,16 @@ def is_match(
                 return True
             except (json.JSONDecodeError, UnicodeDecodeError):
                 data_file.seek(0)
-
+            json_identifier_re = re.compile(r"(:|\[)")
```
Comment (JGSweets) [inline dataprofiler/data_readers/json_data.py:397]: allows both `:` and `[` as a JSON identifier differentiating from just a string.

File: dataprofiler/tests/data_readers/test_json_data.py [hunk 13]
```diff
@@ -12,11 +12,12 @@
 
 class TestNestedJSON(unittest.TestCase):
     def test_flat_to_nested_json(self):
-        dic = {"a.b": "ab", "a.c": "ac", "a.d.f": "adf", "b": "b"}
+        dic = {"a.b": "ab", "a.c": "ac", "a.d.f": "adf", "b": "b", 1: 3}
```
Comment (JGSweets) [inline dataprofiler/tests/data_readers/test_json_data.py:15]: updates test to check for keys that aren't strings

File: dataprofiler/tests/data_readers/test_json_data.py [hunk 14]
```diff
@@ -56,6 +57,11 @@ def setUpClass(cls):
                 encoding="utf-8",
                 count=14,
             ),
+            dict(
```
Comment (JGSweets) [inline dataprofiler/tests/data_readers/test_json_data.py:60]: adds test which includes the new json format

File: dataprofiler/tests/data/json/simple-list.json [hunk 15]
```diff
@@ -0,0 +1,3 @@
+[1]
```
Comment (JGSweets) [inline dataprofiler/tests/data/json/simple-list.json:1]: new data for testing json reading

**Review by ksneab7 — APPROVED** [review]

**Review by taylorfturner — APPROVED** [review]

**Pull request merged**